pageserver/src/page_service.rs

```diff
@@ -435,103 +378,103 @@ impl PageServerHandler {
-                                .observe_closure_duration(|| {
-                                    self.handle_get_nblocks_request(&timeline, &req)
-                                }),
-                            PagestreamFeMessage::GetPage(req) => SMGR_QUERY_TIME
-                                .with_label_values(&["get_page_at_lsn", &tenant_id, &timeline_id])
-                                .observe_closure_duration(|| {
-                                    self.handle_get_page_at_lsn_request(&timeline, &req)
-                                }),
-                            PagestreamFeMessage::DbSize(req) => SMGR_QUERY_TIME
-                                .with_label_values(&["get_db_size", &tenant_id, &timeline_id])
-                                .observe_closure_duration(|| {
-                                    self.handle_db_size_request(&timeline, &req)
-                                }),
-                        };
-
-                        let response = response.unwrap_or_else(|e| {
-                            // print the all details to the log with {:#}, but for the client the
-                            // error message is enough
-                            error!("error reading relation or page version: {:?}", e);
-                            PagestreamBeMessage::Error(PagestreamErrorResponse {
-                                message: e.to_string(),
-                            })
-                        });
-
-                        pgb.write_message(&BeMessage::CopyData(&response.serialize()))?;
-                    } else {
-                        break;
-                    }
+                _ = task_mgr::shutdown_watcher() => {
+                    // We were requested to shut down.
+                    info!("shutdown request received in page handler");
+                    break;
                 }
-                Err(e) => {
-                    if !is_socket_read_timed_out(&e) {
-                        return Err(e);
-                    }
+
+                msg = pgb.read_message() => { msg }
+            };
+
+            let copy_data_bytes = match msg? {
+                Some(FeMessage::CopyData(bytes)) => bytes,
+                Some(m) => {
+                    bail!("unexpected message: {m:?} during COPY");
                 }
-            }
-            drop(profiling_guard);
+                None => break, // client disconnected
+            };
+
+            trace!("query: {:?}", copy_data_bytes);
+
+            let zenith_fe_msg = PagestreamFeMessage::parse(copy_data_bytes)?;
+            let tenant_str = tenant_id.to_string();
+            let timeline_str = timeline_id.to_string();
+
+            let response = match zenith_fe_msg {
+                PagestreamFeMessage::Exists(req) => {
+                    let _timer = SMGR_QUERY_TIME
+                        .with_label_values(&["get_rel_exists", &tenant_str, &timeline_str])
```

Comment:
Nit: with_label_values does some amount of work so may worth to create separate structure like TimelineMetrics which will cache resolved metrics handles so when request arrives we do not need to

I'll open separate PR on top of yours to adjust it